crates/cheatcodes/src/inspector.rs

```diff
@@ -1556,6 +1587,50 @@ impl Cheatcodes {
         }
     }
 
+    #[cold]
+    fn meter_gas_record<DB: DatabaseExt>(
+        &mut self,
+        interpreter: &mut Interpreter,
+        ecx: &mut EvmContext<DB>,
+    ) {
+        if matches!(interpreter.instruction_result, InstructionResult::Continue) {
+            self.gas_metering.gas_records.iter_mut().for_each(|record| {
+                if ecx.journaled_state.depth() == record.depth + 1 {
+                    match interpreter.current_opcode() {
+                        op::CREATE |
+                        op::CALL |
+                        op::CALLCODE |
+                        op::DELEGATECALL |
+                        op::CREATE2 |
+                        op::STATICCALL |
+                        op::EXTSTATICCALL |
+                        op::EXTDELEGATECALL => {
+                            // Reset gas used when entering a new frame.
+                            self.gas_metering.last_gas_used = 0;
+                        }
+                        _ => {}
+                    }
+
+                    // Initialize after new frame, use this as the starting point.
+                    if self.gas_metering.last_gas_used == 0 {
+                        self.gas_metering.last_gas_used = interpreter.gas.spent();
+                        return;
+                    }
```

Comment:
looks like this part conditionally sets `last_gas_used` to 0 and then immediately sets it to `interpreter.gas.spent();`? what do we need this for?

Reply:
If the idea here is to reset the `last_gas_used` to 0 for cases when we track multiple snapshots on different depths, I think this can be done a bit differently by keeping a stack of `last_gas_used` values which we push values to on `step` and pop on `step_end`

Reply:
Refactored this, the resetting to 0 upon encountering an opcode that marks the entry of a new frame is so that the diff will be measured from the first opcode inside of the new frame as otherwise it includes the gas the frame is entered with rather than the gas used by the actual opcode

Without resetting the snapshot diff looks like this

```
[ComparisonGroup] Failed to match snapshots:
- [testGasComparisonCreateA] 99177 → 102352
- [testGasComparisonEmptyA] 0 → 3283
- [testGasComparisonExternalA] 26774 → 197639
- [testGasComparisonInternalColdA] 22100 → 25295
- [testGasComparisonInternalWarmA] 100 → 25445
- [testGasComparisonNestedCallsA] 31953 → 326944
```

Reply:
the code still sets `last_gas_used` to 0 and then immediately to `interpreter.gas.spent()`, I think the first branch does not have effect at all?

I believe when we are entering a new frame, the depth will change and we won't record the steps in it because we are only interested in the depth where snapshot was invoked?

Reply:
you are correct - updated, the setting to `0` wasn't necessary, just the skip over the setting of `record.gas_used` and the update to `gas_metering.last_gas_used` for the next iteration after a encountering a callframe is necessary